### PR TITLE
Goto diff tests and access qualifier comparison

### PR DESCRIPTION
**Pull request by jgwilson42**
Tests for goto-diff. @peterschrammel please could you review? Also, I would be interested in your comments on the test marked as known bug.

### CODE REVIEW

File: regression/goto-diff/java-scope-change2/test.desc [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+old.jar --json-ui
+// Enable multi-line checking
+activate-multi-line-match
+EXIT=0
+SIGNAL=0
+  "deletedFunctions": \[ \],
+  "modifiedFunctions": \[\n    {\n      "name": "java::Test\.foo:\(I\)I",\n      "sourceLocation": {\n        "file": "Test\.java",\n        "function": "java::Test\.foo:\(I\)I",\n        "line": "5"\n      }\n    }\n  \],
+  "newFunctions": \[ \],
+--
+^warning: ignoring
```
Comment (tautschnig) [inline regression/goto-diff/java-scope-change2/test.desc]: Would it be possible to include a comment that says what's currently going wrong here, i.e., which of the expected patterns is not showing up?

Comment (jgwilson42) [reply regression/goto-diff/java-scope-change2/test.desc]: Good point, I have added a couple of comments in the file.

Comment (tautschnig) [reply regression/goto-diff/java-scope-change2/test.desc]: Would you mind moving the comments to the end of the file, after another `--`? That would keep the file syntactically valid (with regard to the syntax expected and documented by the Perl script).

File: regression/goto-diff/java-scope-change/test.desc [hunk 2]
```diff
@@ -0,0 +1,12 @@
+CORE
+new.jar
+old.jar --json-ui
+// Enable multi-line checking
+activate-multi-line-match
+EXIT=0
+SIGNAL=0
+  "deletedFunctions": \[ \],
+  "modifiedFunctions": \[\n    {\n      "name": "java::Test\.foo:\(I\)I",\n      "sourceLocation": {\n        "file": "Test\.java",\n        "function": "java::Test\.foo:\(I\)I",\n        "line": "4"\n      }\n    }\n  \],
```
Comment (peterschrammel) [inline regression/goto-diff/java-scope-change/test.desc]: I'd expect both methods to be modified (both have become non-public). Is the modification in the body of foo intended?

File: regression/goto-diff/java-scope-change2/test.desc [hunk 3]
```diff
@@ -0,0 +1,14 @@
+// Known bug as currently no difference is reported.
+// At least foo should report as changed as the variable within the class has a scope change (public to private).
```
Comment (peterschrammel) [inline regression/goto-diff/java-scope-change2/test.desc]: Access modifiers are not compared at the moment. I'll fix that.

File: regression/goto-diff/java-scope-change/test.desc (hunk 2 above)
Comment (jgwilson42) [reply regression/goto-diff/java-scope-change/test.desc]: Thanks Peter, the modification in the body was a copy and paste error. I have fixed this with the last upload. I have also marked this test as KNOWNBUG given that it now fails.

You are correct that both methods should be marked as modified I have updated the test.desc for this.

File: regression/goto-diff/java-scope-change2/test.desc (hunk 1 above)
Comment (jgwilson42) [reply regression/goto-diff/java-scope-change2/test.desc]: No problem. should be done in the last upload.

File: regression/goto-diff/java-scope-change2/test.desc (hunk 3 above)
Comment (jgwilson42) [reply regression/goto-diff/java-scope-change2/test.desc]: Thanks.

File: regression/goto-diff/java-mod-identifier2/new/Test.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
+public class Test {
+  public int foo(int x) {
+    if (x > 10) {
+      return x;
+    } else {
+      int y = x * 10;
```
Comment (smowton) [inline regression/goto-diff/java-mod-identifier2/new/Test.java]: This has done more than modify an identifier, as it adds an extra statement too. I suggest adding a test like:

```
int f(int x, int y) { return x; }
->
int f(int x, int y) { return y; }
```

That will test whether something that *looks* like a renaming but actually makes a substantial change is detected.

Comment (jgwilson42) [reply regression/goto-diff/java-mod-identifier2/new/Test.java]: Thanks @smowton I have fixed this in the most recent upload. @peterschrammel if you are still working on this branch I have done a force push to put the fix in the original commit.

**Comment by peterschrammel** [comment]
@jgwilson42, do you intend to add more tests here right away? Otherwise, I suggest that we get this merged.

**Comment by jgwilson42** [comment]
@peterschrammel I feel I have enough confidence in the utility to move forward with using it in Deeptest. I suggest we merge this PR and then let the QA Engineers complete the regression suite.

**Review by peterschrammel — APPROVED** [review]

**Comment by peterschrammel** [comment]
@smowton, could you please re-review?

**Review by smowton — APPROVED** [review]

**Pull request merged**